internal/locate/region_cache.go

```diff
@@ -2584,6 +2578,169 @@ func (r *Region) ContainsByEnd(key []byte) bool {
+)
+
+type StoreHealthStatus struct {
+	isSlow atomic.Bool
+
+	// A statistic for counting the request latency to this store
+	clientSideSlowScore SlowScoreStat
+
+	tikvSideSlowScore struct {
+		sync.Mutex
+
+		// The following atomic fields is designed to be able to be read by atomic options directly, but only written
+		// while holding the mutex.
+
+		hasTiKVFeedback atomic.Bool
+		score           atomic.Int64
+		lastUpdateTime  atomic.Pointer[time.Time]
+	}
+}
+
+type HealthStatusDetail struct {
+	ClientSideSlowScore int64
+	TiKVSideSlowScore   int64
+}
+
+func newStoreHealthStatus() *StoreHealthStatus {
+	return &StoreHealthStatus{}
+}
+
+// IsSlow returns whether current Store is slow.
+func (s *StoreHealthStatus) IsSlow() bool {
+	return s.isSlow.Load()
+}
+
+// GetHealthStatusDetail gets the current detailed information about the store's health status.
+func (s *StoreHealthStatus) GetHealthStatusDetail() HealthStatusDetail {
+	return HealthStatusDetail{
+		ClientSideSlowScore: int64(s.clientSideSlowScore.getSlowScore()),
+		TiKVSideSlowScore:   s.tikvSideSlowScore.score.Load(),
+	}
+}
+
+// tick updates the health status that changes over time, such as slow score's decaying, etc. This function is expected
+// to be called periodically.
+func (s *StoreHealthStatus) tick(now time.Time) {
+	s.clientSideSlowScore.updateSlowScore()
+	s.updateTiKVServerSideSlowScoreOnTick(now)
+	s.updateSlowFlag()
+}
+
+// recordClientSideSlowScoreStat records timecost of each request to update the client side slow score.
+func (s *StoreHealthStatus) recordClientSideSlowScoreStat(timecost time.Duration) {
+	s.clientSideSlowScore.recordSlowScoreStat(timecost)
+	s.updateSlowFlag()
+}
+
+// markAlreadySlow marks the related store already slow.
+func (s *StoreHealthStatus) markAlreadySlow() {
+	s.clientSideSlowScore.markAlreadySlow()
+	s.updateSlowFlag()
+}
+
+// updateTiKVServerSideSlowScoreOnTick updates the slow score actively, which is expected to be a periodic job.
+// It skips updating if the last update time didn't elapse long enough, or it's being updated concurrently.
+func (s *StoreHealthStatus) updateTiKVServerSideSlowScoreOnTick(now time.Time) {
+	if !s.tikvSideSlowScore.hasTiKVFeedback.Load() {
+		// Do nothing if no feedback has been received from this store yet.
+		return
+	}
+	lastUpdateTime := s.tikvSideSlowScore.lastUpdateTime.Load()
+	if lastUpdateTime == nil || now.Sub(*lastUpdateTime) < tikvSlowScoreUpdateFromPDInterval {
+		// If the first feedback is
+		return
+	}
+
+	if !s.tikvSideSlowScore.TryLock() {
+		// It must be being updated concurrently.
+		return
+	}
+	defer s.tikvSideSlowScore.Unlock()
+
+	// Reload update time as it might be updated concurrently before acquiring mutex
+	lastUpdateTime = s.tikvSideSlowScore.lastUpdateTime.Load()
+	elapsed := now.Sub(*lastUpdateTime)
+	if elapsed < tikvSlowScoreUpdateFromPDInterval {
+		return
+	}
+
+	// TODO: Try to get store status from PD here. But it's not mandatory.
```

Comment:
The function would be executed inside the `batchRecvLoop`, would it affect the performance if we call `pdClient.func` here?

Reply:
No. This function is not expected to be run in batchRecvLoop, but should be a periodic task executed in the background. 
 The function that might be called inside batchRecvLoop is `updateTiKVServerSideSlowScore`.

Here what we need to be careful about is the mutex.

Reply:
Got it, it's better to add comments to functions that would be used in the critical performance path.